### DISCUSSION TITLE
ImageSharp (2.0+) and OpenGL

### DISCUSSION THREAD

**Discussion by ggolda** [SixLabors/ImageSharp#2499] [Q&A]
Hello again!
Recently got back to my old project that used ImageSharp 1.0 and decided to update to the newest version (3.0) and found out that it has drastically changed it's API and how images are represented in memory. And currently I don't see how to glue it together with OpenGL.

So in previous version to get a pointer to underlying block of memory and load it to OpenGL texture, I used code like that:
```c#
        private unsafe void LoadTexture(Image<Rgba32> texture)
        {
            texture.TryGetSinglePixelSpan(out var pixels);
            var bytes = MemoryMarshal.AsBytes(pixels);
            fixed (Rgba32* pointer = &MemoryMarshal.GetReference(pixels))
            {
                GL.TexSubImage2D(TextureTarget.Texture2D, 0, 0, 0, _width, _height, PixelFormat.Rgba, PixelType.UnsignedByte, (IntPtr)pointer);
            }
        }
```

As I understand, it's no longer possible to extract a span or pointer to pixels without copying to some intermediate array?

**Comment by antonfirsov** [comment]
Since 2.0+ image buffers are not guaranteed to be contiguous by default, you need to enforce it:
https://docs.sixlabors.com/articles/imagesharp/memorymanagement.html#enforcing-contiguous-buffers
